File: src/C++/rank_trace_matrix.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-23 17:39:22
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-23 19:38:54
@FilePath: \asm_matrix_benchmark\src\C++\rank_trace_matrix.cpp
@Description: rank and trace matrix c++ code
*************************************************************/
#include "matrix_cpp.h"
#include <algorithm>
#include <cmath>

/***********************************************************
@description: calc rank of matrix
@return {*}
*************************************************************/
size_t MatrixInt::rank() const
{
    if (rows == 0 || cols == 0)
    {
        return 0;
    }

    // copy to 2D vector for elimination
    std::vector<std::vector<double>> mat(rows);
    for (size_t i = 0; i < rows; i++)
    {
        mat[i].resize(cols);
        for (size_t j = 0; j < cols; j++)
        {
            mat[i][j] = static_cast<double>(data[i * cols + j]);
        }
    }

    size_t r = 0;
    for (size_t col = 0; col < cols && r < rows; col++)
    {
        size_t pivot = r;
        while (pivot < rows && std::fabs(mat[pivot][col]) < 1e-12)
        {
            pivot++;
        }
        if (pivot == rows)
        {
            continue;
        }

        std::swap(mat[r], mat[pivot]);

        for (size_t i = r + 1; i < rows; i++)
        {
            if (std::fabs(mat[i][col]) >= 1e-12)
            {
                double factor = mat[i][col] / mat[r][col];
                for (size_t j = col; j < cols; j++)
                {
                    mat[i][j] -= factor * mat[r][j];
                }
            }
        }
        r++;
    }
    return r;
}

/***********************************************************
@description: calc trace of matrix
@return {*}
*************************************************************/
int MatrixInt::trace() const
{
    if (rows != cols)
    {
        throw std::invalid_argument("trace: square matrix required");
    }
    int sum = 0;
    for (size_t i = 0; i < rows; i++)
    {
        sum += data[i * cols + i];
    }
    return sum;
}

File: src/JavaScript/rank_trace_matrix.js
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-29 20:10:32
@LastEditors: GitHubCopilot
@LastEditTime: 2026-06-29 20:10:32
@FilePath: \asm_matrix_benchmark\src\JavaScript\rank_trace_matrix.js
@Description: rank and trace matrix JavaScript code
*************************************************************/

"use strict";

const {
  checkMatrix,
  gaussianRank,
  EPS_FLOAT,
  EPS_DOUBLE,
  castByType,
} = require("./common");

/**
 * Compute matrix rank under numeric tolerance.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {"int"|"float"|"double"} type
 * @returns {number|null}
 */
function rankMatrix(m, type) {
  if (!checkMatrix(m, "m")) {
    return null;
  }

  let eps = 0;
  if (type === "float") {
    eps = EPS_FLOAT;
  } else if (type === "double") {
    eps = EPS_DOUBLE;
  }

  return gaussianRank(m, eps);
}

/**
 * Compute matrix trace. Matrix must be square.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {"int"|"float"|"double"} type
 * @returns {number|null}
 */
function traceMatrix(m, type) {
  if (!checkMatrix(m, "m")) {
    return null;
  }
  if (m.rows !== m.cols) {
    console.error("Matrix must be square!");
    return null;
  }

  let sum = 0;
  for (let i = 0; i < m.rows; i++) {
    sum += m.data[i * m.cols + i];
  }
  return castByType(sum, type);
}

function rank_matrix_int(m) {
  return rankMatrix(m, "int");
}

function trace_matrix_int(m) {
  return traceMatrix(m, "int");
}

function rank_matrix_float(m) {
  return rankMatrix(m, "float");
}

function trace_matrix_float(m) {
  return traceMatrix(m, "float");
}

function rank_matrix_double(m) {
  return rankMatrix(m, "double");
}

function trace_matrix_double(m) {
  return traceMatrix(m, "double");
}

module.exports = {
  rank_matrix_int,
  trace_matrix_int,
  rank_matrix_float,
  trace_matrix_float,
  rank_matrix_double,
  trace_matrix_double,
};
